### ISSUE TITLE
Rule request: no-useless-return

### ISSUE THREAD

**Issue by Wilfred** [eslint/eslint#7309]
**Please describe what the rule should do:**

Forbid `return;` at the end of a function, as it is redundant. It's often a leftover from refactoring some callback code.

**What category of rule is this? (place an "X" next to just one item)**

[x] Enforces code style
[ ] Warns about a potential error
[ ] Suggests an alternate way of doing something
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

The common case:

```js
function foo() {
  bar();
  return; // pointless!
}
```

Generalising to conditionals at the end of a function:

```js
function foo() {
  if (something) {
    bar();
    return; // redundant
  } else if (otherThing) {
    baz();
    return; // redundant
  } else {
    boz();
  }
  // Nothing here, so the arms of the conditional are the last thing
  // in the function.
}
```

Examples where we wouldn't warn:

```js
function foo() {
  bar();
  return 1;
}
```

**Why should this rule be included in ESLint (instead of a plugin)?**

It's totally generic and applicable to all JS developers I think.

**Comment by platinumazure** [comment]
@Wilfred Thanks for following our new rule proposal guidelines.

I've got just a couple of questions:

1. Should return statements with an expression be warned? (I assume no)
1. Should `return void 0` (or other uses of void operator) be warned?

Also, are you willing to create this rule yourself?

**Comment by Wilfred** [comment]
1. Return statements with an expression should not be warned -- I've updated the issue to reflect this.
2. I think `return void 0` should not be warned either -- that's probably a sign of the user doing it deliberately. There's already `no-void` and `consistent-return`.

I'd be willing to write the rule, but I may need a few pointers/patient reviewers.

**Comment by platinumazure** [comment]
Sounds great! I'll endorse this. With two more endorsements and a champion, we can accept this.

**Comment by Wilfred** [comment]
I've got three thumbs-up on my issue -- are those equivalent to endorsements?

**Comment by platinumazure** [comment]
@Wilfred Only thumbs-up from the core team count as endorsements in this case-- we're at 2 right now. (I thought we had documented this in the new rule/rule change processes, but I guess we don't-- see [here](http://eslint.org/docs/maintainer-guide/issues.html#consensus) for more info. Sorry for any confusion.)

**Comment by nzakas** [comment]
@platinumazure it's probably best not to use the terms "endorse" or "endorsement", as even I was confused about what that meant. Let's just stick with saying you've got 3 👍s.

I'd suggest simplifying the wording a bit on this proposal to fit with our other rules: either `no-useless-return` or `no-extra-return`.

We still need a champion for this proposal.

**Comment by not-an-aardvark** [comment]
I'll champion this.

**Comment by platinumazure** [comment]
@nzakas I had been using the term for over a month with no complaint, so I guess I misconstrued silence for acceptance. Do we have an official term for when the graphic isn't available?

**Comment by Wilfred** [comment]
I've updated the issue to reflect the new name.

**Comment by not-an-aardvark** [comment]
@Wilfred, have you begun implementing this? (I ask because I'm willing to start implementing this, but I don't want to duplicate effort if you're already working on it.)

**Comment by Wilfred** [comment]
No, I haven't begun implementing this. If you fancy coding it, feel free, but I'm happy to code it up too. Let me know :)

**Comment by mysticatea** [comment]
I can implement this as well since I like code path analysis. :smile:

**Comment by not-an-aardvark** [comment]
@mysticatea Feel free to work on it, I was having some trouble with code path analysis.

I pushed a branch [here](https://github.com/eslint/eslint/compare/no-useless-return-7309) that has tests and docs for the the rule already, but the implementation doesn't work yet.

**Comment by mysticatea** [comment]
@not-an-aardvark Thank you!
